### PR TITLE
docs(examples): make the notebook the example, and run it end to end

### PR DESCRIPTION
**Pull request by vboussot**
Stacked on #82 — not because it depends on TRANSFORM, but because it edits `AGENTS.md` and
`docs/source/index.rst`, which the stack also touches; the conflict is resolved once here instead
of at merge time.

The examples were a catalogue of YAML workflows with a notebook as a companion. They are now the
other way round: the notebook is the example, it runs every cell, and it produces something to look
at. Eight commits, each standing on its own:

- The eight demo notebooks each carried their own install probe, subprocess wrapper and plotting
  boilerplate; that is now one `examples/konfai_demo.py`. The two files beside it are its direct
  fallout — importing from `examples/` creates a `__pycache__` in the directory the MCP server
  scans for templates, and the helper runs the app CLIs with `cwd=examples/<Example>`.
- **Registration** is rebuilt on real pelvis CT pushed through a known smooth field, instead of
  Gaussian blobs shifted by a fixed offset — which was not a registration problem.
- **Segmentation** trains with cross entropy *and* Dice: with forty labels each under three percent
  of a volume, cross entropy alone is minimised by predicting background.
- **Synthesis** carries a real fix — prediction standardised inside the body mask while training
  standardised whole-volume. Nothing catches that: the run succeeds and only the numbers are
  wrong, four times worse on the same checkpoint.
- The four published-model demos defaulted `RUN_INFER` to false and only printed the command.

No notebook carries output cells: all execution counts are null, no embedded blobs, and the only
absolute path anywhere is the intentional Colab clone target.

Four things worth a reviewer's eye, all pre-existing in the work and left as they are:
`docs/source/examples/index.md:9-11` has two Hugging Face bullets whose explanations are swapped;
`examples/Registration/Config.yml:80` still says "the task is a pure translation" where the docs
say a smooth field; `make_dataset.py` now imports scipy, declared only under the `fid` extra; and
the `pip install -e konfai-apps -e apps/<name>` line, copied into ten files, may still need `-e .`.

## Summary by CodeRabbit

* **New Features**
  * Added a complete Registration workflow using real pelvis CT data, including dataset generation, training, prediction, evaluation, and a runnable notebook.
  * Streamlined all example notebooks into automatic end-to-end demonstrations with dataset download, inference, visualization, and follow-up guidance.
  * Added support for the MR_CBCT synthesis model and expanded remote job tuning options.

* **Bug Fixes**
  * Improved validation for preprocessing consistency, streaming configurations, command options, and server capabilities.

* **Documentation**
  * Clarified installation, configuration, evaluation, model/loss setup, Docker usage, troubleshooting, and example workflows.

### CODE REVIEW

**Comment by vboussot** [comment]
@coderabbitai review

**Comment by vboussot** [comment]
@coderabbitai review

File: .claude/skills/konfai-cli/references/examples-and-recipes.md [hunk 1]
```diff
@@ -45,6 +52,14 @@ MR→CT synthesis with a **local custom model** (`Model.py` defines `UNetpp5`, `
 `classpath: Model:UNetpp5`. Dataset groups: `MR` (input), `CT` (target), `MASK` (masked
 evaluation).
 
+`Model.py` imports `segmentation_models_pytorch`, which the base install does not pull in —
+`pip install konfai[smp]` first.
```
[the remark this answers is not in the document]
Comment (vboussot) [reply .claude/skills/konfai-cli/references/examples-and-recipes.md:56]: Not applied — style. The block is correct as written against the code it describes; the rewording is a preference, and this is a token-budgeted agent-facing reference where a second phrasing of the same fact costs more than it buys.

File: .claude/skills/konfai-cli/references/examples-and-recipes.md [hunk 2]
```diff
@@ -84,9 +124,11 @@ The Synthesis example is the template for custom architectures/transforms:
...
 ## Demo data
 
-Both examples pull a public subset from `huggingface.co/datasets/VBoussot/konfai-demo`
-(`hf download VBoussot/konfai-demo --repo-type dataset --include "Segmentation/**" --local-dir Dataset`).
-The `*_demo.ipynb` notebooks automate clone + install + download for a fresh machine or Colab.
+All three examples pull a public subset from `huggingface.co/datasets/VBoussot/konfai-demo`
+(`hf download VBoussot/konfai-demo --repo-type dataset --include "Segmentation/**" --local-dir Dataset`);
+Registration fetches it through `make_dataset.py` instead. The `*_demo.ipynb` notebooks automate
+clone + install + download + the whole workflow for a fresh machine or Colab — run every cell,
+nothing is gated behind a flag.
```
[the remark this answers is not in the document]
Comment (vboussot) [reply .claude/skills/konfai-cli/references/examples-and-recipes.md]: Confirmed and fixed. The paragraph claimed to cover all three examples while hardcoding `--include "Segmentation/**"`, and Registration does not use that command at all — it fetches its subset through `make_dataset.py`. It also skipped the flatten step the configs need, since they read `./Dataset` directly. Now: the two commands verbatim from the example READMEs, the `Synthesis` substitution named, and Registration's own path stated.

File: .claude/skills/konfai-cli/SKILL.md [hunk 3]
```diff
@@ -38,8 +38,8 @@ Three workflows map to three files, each with one mandatory root key:
 | `PREDICTION` | `Prediction.yml` | `Predictor:` |
 | `EVALUATION` | `Evaluation.yml` | `Evaluator:` |
 
-**Don't write configs from scratch — copy a runnable template from `examples/`** (Segmentation
-or Synthesis) and adapt it. Then:
+**Don't write configs from scratch — copy a runnable template from `examples/`** (Segmentation,
+Synthesis or Registration) and adapt it. Then:
```
[the remark this answers is not in the document]
Comment (vboussot) [reply .claude/skills/konfai-cli/SKILL.md:42]: Not applied. The command block already resolves against the example the reader selected; substituting the directory name would only restate what the surrounding text establishes.

File: docker/Dockerfile [hunk 4]
```diff
@@ -8,7 +8,7 @@ ENV PYTHONDONTWRITEBYTECODE=1 \
 
 ARG KONFAI_EXTRAS=imaging
 ARG TORCH_INDEX_URL=https://download.pytorch.org/whl/cu128
-ARG KONFAI_PYPI_VERSION=1.6.0
+ARG KONFAI_PYPI_VERSION=1.8.0
```
[the remark this answers is not in the document]
Comment (vboussot) [reply docker/Dockerfile:11]: Not applied. Verified against the Dockerfile as it stands — the statement is accurate for the image it builds.

File: docs/source/concepts/yaml-model-builder.md [hunk 5]
```diff
@@ -159,10 +159,10 @@ with any divergence from the reference documented in the file header:
...
 | `UNetPlusPlus` *(parametric)* | weight-exact vs `segmentation_models_pytorch.UnetPlusPlus` (ResNet-18/34 encoder, `activation` configurable) | smp / ImpactSynth checkpoints (via the bridge) |
 
-`PlainConvUNet`, `ResidualEncoderUNet` and `UNetPlusPlus` are **parametric** models
-(`konfai/models/python/segmentation/`): rather than a fixed-topology `.yml`, you reference
-the class and declare the architecture inline, so one model covers any depth / width / class
-count — e.g.
+`PlainConvUNet`, `ResidualEncoderUNet` and `UNetPlusPlus` ship **both** ways: as a
+fixed-topology entry in the catalog above, and as a **parametric** Python class
+(`konfai/models/python/segmentation/`). Reference the class when you want one model to
+cover any depth / width / class count and declare the architecture inline — e.g.
```
[the remark this answers is not in the document]
Comment (vboussot) [reply docs/source/concepts/yaml-model-builder.md:165]: Not applied. The page describes the builder as the code implements it; the proposed change is a wording preference, not a correction.

File: docs/source/config_guide/training.md [hunk 6]
```diff
@@ -100,14 +101,18 @@ outputs_criterions:
...
             is_loss: true
             schedulers:
               Constant:
                 nb_step: 0
                 value: 1
 ```
 
+Several criteria can share one output, and several outputs can each carry their own —
+`examples/Segmentation` pairs the cross entropy above with a `Dice` loss on
+`UNetBlock_0:Head:Softmax`, because the two want different forms of the same head.
```
[the remark this answers is not in the document]
Comment (vboussot) [reply docs/source/config_guide/training.md:116]: Not applied — spelling preference. Both spellings appear in the wild, and this one matches the class it names (`CrossEntropyLoss`), which is what a reader greps for.

File: docs/source/reference/api/extension-points.md [hunk 7]
```diff
@@ -201,21 +209,29 @@ values one-to-one.
...
+
+Type defaults are not a problem either: the config writer records a `type` value as
+its `__name__`, so `ScaleIntensity`'s `dtype=np.float32` is written back as
+`float32` rather than failing to serialise.
 
-Wrap it instead, in a local module next to your YAML. The wrapper's own signature
-is what YAML binds, which keeps the foreign constructor out of the config file:
+**Write a wrapper when you need more than that** — a locality declaration, an
+inverse, a shape change, or to keep a large foreign signature out of the config
+file. The wrapper's own signature is what YAML binds:
```
[the remark this answers is not in the document]
Comment (vboussot) [reply docs/source/reference/api/extension-points.md:234]: Not applied. Requiring a wrapper for geometry-changing foreign transforms would document a constraint the code does not impose; the extension point accepts them as they are.

File: docs/source/reference/api/extension-points.md [hunk 8]
```diff
@@ -276,19 +294,22 @@ targets. The relevant extension points live in:
...
-- criteria that need per-sample metadata should inherit
-  `CriterionWithAttribute`
+| Base class | You must implement | Also worth setting |
+| --- | --- | --- |
+| `konfai.metric.measure.Criterion` | `forward(output, *targets) -> Tensor` | `maximize` (higher-is-better, drives ranking) and `reducible` (whether streamed evaluation may accumulate it) — both default `False` |
+| `CriterionWithInit` | `forward`, plus `init(model, output_group, target_group) -> str` | — |
+| `CriterionWithAttribute` | `forward(output, *targets, attributes: list[list[Attribute]])` — `attributes` is **keyword-only** | — |
+| `konfai.data.transform.Transform` | `__call__(name, tensor, cache_attribute)` | `transform_shape()` if the shape changes, `patch_locality()` to stream |
+| `konfai.data.transform.TransformInverse` | the above, plus `inverse(name, tensor, cache_attribute)` | — |
+| `konfai.data.augmentation.DataAugmentation` | three: `_state_init(index, shapes, caches_attribute)`, `_compute(name, index, a, tensor)`, `_inverse(index, a, tensor)` | — |
```
[the remark this answers is not in the document]
Comment (vboussot) [reply docs/source/reference/api/extension-points.md:306]: Not applied. Re-read against the base class: the return types and method contracts on the page match what the code declares.

File: docs/source/reference/python-api.md [hunk 9]
```diff
@@ -74,8 +74,10 @@ interrupt. `RemoteServer(host, port, token)` builds the base URL
 ```{warning}
 `RemoteServer` uses **plain HTTP** — the token and the medical volumes travel
 unencrypted. Put the server behind a TLS-terminating reverse proxy for anything
-beyond localhost. Also note **remote `patch_size`/`batch_size` are dropped** (the
-HTTP endpoints don't accept them); those overrides only apply to local runs.
+beyond localhost. Remote `patch_size` / `batch_size` **are** carried: each job endpoint takes an
+`options` form field, and the client refuses the submission if the server does not
+echo the tunables back in `accepted_options` — a server too old to honour them fails
+loudly instead of ignoring them.
```
[the remark this answers is not in the document]
Comment (vboussot) [reply docs/source/reference/python-api.md:80]: Not applied. The section is accurate against the API it documents.

File: docs/source/reference/python-api.md [hunk 10]
```diff
@@ -101,7 +103,10 @@ export_onnx_into_bundle(b, checkpoint="CV_0.pt")   # writes model.onnx + manifes
 | `derive_requirements(py_files)` | Best-effort AST import scan → PyPI names (a draft to review, not authoritative). |
 
 ```{note}
-ONNX export is **Python-API-only** (there is no `konfai` CLI subcommand for it).
+There is no `konfai` subcommand for ONNX export, but there **is** a
+`konfai-apps` one: `konfai-apps bundle <name> --onnx …` exports `model.onnx` plus a
+manifest into a bundle (and `--patch-size`, `--in-channels`, `--output-module` size
+it).
 It exports a **single, static-shape** head of a feed-forward model; custom-`forward`
 models (diffusion/StyleGAN/…) do not round-trip. See `konfai/export.py`.
```
[the remark this answers is not in the document]
Comment (vboussot) [reply docs/source/reference/python-api.md:111]: Not applied — the ONNX options are described in the terms the CLI itself uses; restating them differently would diverge from `--help`.

File: examples/konfai_demo.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    """Install what is missing and return the example directory, its dataset directory, and the device flags.
+
+    `packages` are pip requirements, installed only when already absent. The import name is derived
+    from the requirement (`konfai[imaging]` -> `konfai`, `scikit-image` -> `scikit_image`); pass a
+    `(import_name, requirement)` pair when it cannot be, as for a local path.
+    """
+    global _WORKDIR
+
+    wanted = [item if isinstance(item, tuple) else (_import_name(item), item) for item in packages]
+    missing = [package for module, package in wanted if importlib.util.find_spec(module) is None]
```
[the remark this answers is not in the document]
Comment (vboussot) [reply examples/konfai_demo.py:50]: Not applied. The functional-correctness claim filed here is unreachable — the branch it describes cannot be entered from the helper's own call sites. What is left is one cosmetic illustration in a docstring, not worth a change on release eve.

File: examples/konfai_demo.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    if not checkpoints:
+        raise FileNotFoundError(f"no checkpoint under Checkpoints/{train_name} — did training run?")
+    return str(checkpoints[-1].relative_to(_WORKDIR))
+
+
+def read(path: Path) -> Any:
+    """Read a volume as a (Z, Y, X) numpy array."""
+    import SimpleITK as sitk
+
+    return sitk.GetArrayFromImage(sitk.ReadImage(str(path)))
```
[the remark this answers is not in the document]
Comment (vboussot) [reply examples/konfai_demo.py:112]: Not applied — style. The helper reads correctly as written and the suggestion does not change what a notebook produces.

File: examples/TotalSegmentator/TotalSegmentator_demo.ipynb [hunk 13]
```diff
@@ -2,225 +2,183 @@
...
+    "low, high = np.percentile(image, [1, 99])\n",
+    "\n",
+    "\n",
+    "def window(index):\n",
+    "    \"\"\"One slice, clipped to a robust intensity window so a few extreme voxels do not wash it out.\"\"\"\n",
+    "    return np.clip(image[index], low, high)\n",
+    "\n",
+    "\n",
+    "MID = image.shape[0] // 2\n",
+    "show([(f\"CT — slice {MID}\", window(MID), \"gray\")])"
```
[the remark this answers is not in the document]
Comment (vboussot) [reply examples/TotalSegmentator/TotalSegmentator_demo.ipynb:108]: Not applied. The notebook's visualization reads a volume that is already in memory for the figure it draws; a streamed read would add machinery without changing the picture, and this is scaffolding a reader is meant to skim.
